circuit_knitting/utils/equivalence.py

```diff
@@ -0,0 +1,133 @@
+# This code is a Qiskit project.
+
+# (C) Copyright IBM 2024.
+
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Equivalence utilities.
+
+.. currentmodule:: circuit_knitting.utils.equivalence
+
+.. autosummary::
+   :toctree: ../stubs/
+
+"""
+from collections import defaultdict
+
+import numpy as np
+from qiskit.circuit import (
+    EquivalenceLibrary,
+    QuantumCircuit,
+    QuantumRegister,
+    Parameter,
+)
+from qiskit.circuit.library.standard_gates import (
+    RZGate,
+    XGate,
+    YGate,
+    ZGate,
+    HGate,
+    SGate,
+    IGate,
+    SdgGate,
+    SXGate,
+    SXdgGate,
+    TGate,
+    TdgGate,
+    RXGate,
+    RYGate,
+    PhaseGate,
+)
+
+_eagle_sel = HeronEquivalenceLibrary = EagleEquivalenceLibrary = EquivalenceLibrary()
+equivalence_libraries = defaultdict(
+    lambda: None, {"heron": EagleEquivalenceLibrary, "eagle": EagleEquivalenceLibrary}
+)
+
+######################################################################
+
+# YGate
+q = QuantumRegister(1, "q")
+def_y = QuantumCircuit(q)
+for inst in [RZGate(np.pi), XGate()]:
+    def_y.append(inst, [0], [])
+_eagle_sel.add_equivalence(YGate(), def_y)
+
+# ZGate
+q = QuantumRegister(1, "q")
+def_z = QuantumCircuit(q)
+def_z.append(RZGate(np.pi), [0], [])
+_eagle_sel.add_equivalence(ZGate(), def_z)
+
+# HGate
+q = QuantumRegister(1, "q")
+def_h = QuantumCircuit(q)
+for inst in [RZGate(np.pi / 2), SXGate(), RZGate(np.pi / 2)]:
+    def_h.append(inst, [0], [])
+_eagle_sel.add_equivalence(HGate(), def_h)
+
+# SGate
+q = QuantumRegister(1, "q")
+def_s = QuantumCircuit(q)
+def_s.append(RZGate(np.pi / 2), [0], [])
+_eagle_sel.add_equivalence(SGate(), def_s)
+
+# SdgGate
+q = QuantumRegister(1, "q")
+def_sdg = QuantumCircuit(q)
+def_sdg.append(RZGate(-np.pi / 2), [0], [])
+_eagle_sel.add_equivalence(SdgGate(), def_sdg)
+
+# SXdgGate
+q = QuantumRegister(1, "q")
+def_sxdg = QuantumCircuit(q)
+for inst in [
+    RZGate(np.pi / 2),
+    RZGate(np.pi / 2),
+    SXGate(),
+    RZGate(np.pi / 2),
+    RZGate(np.pi / 2),
+]:
+    def_sxdg.append(inst, [0], [])
+_eagle_sel.add_equivalence(SXdgGate(), def_sxdg)
+
+# TGate
+q = QuantumRegister(1, "q")
+def_t = QuantumCircuit(q)
+def_t.append(RZGate(np.pi / 4), [0], [])
+_eagle_sel.add_equivalence(TGate(), def_t)
+
+# TdgGate
+q = QuantumRegister(1, "q")
+def_tdg = QuantumCircuit(q)
+def_tdg.append(RZGate(-np.pi / 4), [0], [])
+_eagle_sel.add_equivalence(TdgGate(), def_tdg)
+
+# RXGate
+q = QuantumRegister(1, "q")
+def_rx = QuantumCircuit(q)
+theta = Parameter("theta")
+for inst in [RZGate(np.pi / 2), SXGate(), RZGate(theta + np.pi), RZGate(5 * np.pi / 2)]:
```

Comment:
I used Qiskit to find these translations, and I didn't change anything, even if it looked suspicious. I'll label the ones I think can be simplified here, and we can discuss. This one appears to rotate 5/2*pi times at the end. I believe we can just rotate 1/2*pi?

Reply:
@jakelishman , we should be able to simplify these strange rotation angles, right? I know the translation process uses a shortest-path algorithm, so I assume it can find paths containing these funky angles. Just wanted to be sure there wasn't something I was missing

Reply:
Since you're acting symbolically here, you can only really use the `BasisTranslator`, which acts on a gate-by-gate basis, which just dumbly multiplies/divides parameters as appropriate during conversion.  We don't do symbolic peephole optimisation (though we could), and we can't do full unitary resynthesis with symbolic parameters, so you'll end up with things like this trying to decompose symbolically in general.

In this particular case, we clearly could add a specific `rx -> [rz,sx]` decomposition if we so chose (at the moment the default equivalence library does some funny roundabout thing).  There's probably lots of places where we could improve the standard equivalence library for symbolic gates - we mostly just decompose to Euler angles and resynthesise into something more efficient during an optimisation pass.